src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs

```diff
@@ -65,14 +67,23 @@ public async virtual Task<HostSecretsInfo> GetHostSecretsAsync()
         {
             if (_hostSecrets == null)
             {
-                HostSecrets hostSecrets = await LoadSecretsAsync<HostSecrets>();
+                HostSecrets hostSecrets;
```

Comment:
Don't we have the same pattern going on in GetFunctionsSecrets below that we need to protect as well?

Reply:
I applied same pattern to GetFunctionsSecrets.

Reply:
That’s pretty aggressive and will need to be revisited. Each function has its own resource for secrets, we’ll be causing unnecessary contention here in cases where secrets are being set for multiple functions

Reply:
Well, concurrent calls to get secrets for the same function would run into the original bug, right? So serializing the writes here as well seems to be required if you want to ensure no failures.

Reply:
Yes, but locking on a single resource for all functions is just inefficient. Ideally, this would address that case you mentioned. With the current implementation, concurrent requests for multiple functions will all lock on the same resource here, with a a storage call for each of them, this would be pretty significant if we're dealing with multiple functions.

Reply:
To be clear, I agree that locking on this pack is required. There isn't a rush to get this in, so we should do it properly so it isn't something we need to come back to.